test/hotspot/jtreg/compiler/predicates/TestHoistedPredicateForNonRangeCheck.java

```diff
@@ -0,0 +1,125 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8307683
+ * @library /test/lib /
+ * @requires vm.compiler2.enabled
+ * @summary Tests that IfNode is not wrongly chosen as range check by Loop Predication leading to crashes and wrong executions.
+ * @run main/othervm -Xcomp -XX:CompileCommand=compileonly,compiler.predicates.TestHoistedPredicateForNonRangeCheck::test
+ *                   compiler.predicates.TestHoistedPredicateForNonRangeCheck
+ * @run main/othervm -Xcomp -XX:CompileCommand=compileonly,compiler.predicates.TestHoistedPredicateForNonRangeCheck::test
+ *                   -XX:LoopMaxUnroll=0 compiler.predicates.TestHoistedPredicateForNonRangeCheck
+ */
+
+/*
+ * @test
+ * @bug 8307683
+ * @library /test/lib /
+ * @summary Tests that IfNode is not wrongly chosen as range check by Loop Predication leading to crashes and wrong executions.
+ * @run main/othervm -Xbatch compiler.predicates.TestHoistedPredicateForNonRangeCheck calendar
+ */
+
+package compiler.predicates;
+
+import jdk.test.lib.Asserts;
+
+import java.util.Calendar;
+import java.util.Date;
+
+
+public class TestHoistedPredicateForNonRangeCheck {
+    static int iFld, iFld2;
+
+    public static void main(String[] args) {
+        if (args.length == 0) {
+            Integer.compareUnsigned(34, 34); // Ensure Integer class is loaded and we do not emit a trap inside test() for it.
+
+            for (int i = 0; i < 2; i++) {
+                iFld = 0;
+                iFld2 = 0;
+                test();
+                Asserts.assertEQ(iFld, 901, "wrong value");
+                Asserts.assertEQ(iFld2, 100, "wrong value");
+            }
+        } else {
+            boolean flag = false;
+            for (int i = 0; i < 10000; i++) {
+                testCalendar1();
+                testCalendar2(flag);
+            }
+        }
+    }
+
+    public static void test() {
+        // (Inverted) Hoisted Predicate checks lower and upper bound: -1 < 0 && 1000 >= 100 -> always true and the predicate is removed
+        // while template assertion predicates are kept. When splitting this loop further, we insert an assertion predicate which
+        // fails for i = 0 and we halt. When not splitting this loop (with LoopMaxUnroll=0), we have a wrong execution due
+        // to never executing iFld2++ (we removed the check and the branch with the trap).
+        for (int i = -1; i < 1000; i++) {
+            if (Integer.compareUnsigned(i, 100) < 0) { // Loop Predication creates a Hoisted Range Check Predicate due to trap with Float.isNan().
```

Comment:
Maybe add a comment explaining that this is equivalent to `i <u 100` <=> `i >= 0 && i < 100` and we add a predicate for the else branch, i.e. for `i < 0 || i >= 100`, and remove the if branch.